WIP Accelerate enabled open source models

thepok opened the pull request:
Many TODOs but already workable.

Now its easy to load big open source models.

I was able to load neox 20b and facebook 30b on my laptop with 64gb ram and 16gb vram.

thepok committed:
WIP Accelerte enabled models

thepok committed:
WIP Accelerte enabled models

thepok committed:
fixing type errors

thepok committed:
fixing type errors

thepok committed:
lint type fix

thepok committed:
lint type fix

thepok committed:
black ...

thepok committed:
black ...

thepok committed:
sort imports

thepok committed:
sort imports

thepok committed:
sort 2

thepok committed:
sort 2

hwchase17 reviewed (commented):
nice! im not super familiar with accerlated stuff. how generic is the get_accelerated_gpt code?

thepok committed:
Merge branch 'hwchase17:master' into accelerate

hwchase17 committed:
Add BaseCallbackHandler and CallbackManager (#478)

Co-authored-by: Ankush Gola <9536492+agola11@users.noreply.github.com>

hwchase17 committed:
bump version to 0056 (#533)

eltociear committed:
Fix typo in key_concepts.md (#535)

therefor -> therefore

rubensmau committed:
Updated embeddings.ipynb (#531)

updated embeddings.ipynb

Yongtae723 committed:
add logger (#529)

As talking #519, I made 2 PRs.

this is the first PR for adding a logger.

I am concerned about the following two points and would appreciate your
opinion.

1. Since the logger is not formatted, the statement itself is output
like a print statement, and I thought it was difficult to understand
that it was a warning, so I put WARNING! at the beginning of the warning
statement. After the logger formatting is done properly, the word
WARNING can be repeated.
2. Statement `Please confirm that {field_name} is what you intended.`
can be replaced like `If {field_name} is intended parameters, enter it
to model_kwargs`
thank you!

Yongtae

Yongtae723 committed:
replace forbid into ignore (#539)

this is the second PR of #519.
in #519 I suggested deleting Extra.forbid.
I was very confused but I replaced Extra.forbid to Extra.ignore, which
is the default of pydantic.


Since the
[BaseLLM](https://github.com/hwchase17/langchain/blob/4b7b8229de50e5e97a2b2c12e1607f347d8ded35/langchain/llms/base.py#L20)
from which it is inherited is set in Extra.forbid, I wanted to avoid
having the Extra.forbid settings inherited by simply deleting it.

samching committed:
Add info to Contributors.md to avoid Conda/Pyenv dependency conflicts  (#540)

As discussed in the
[Discord](https://discord.com/channels/1038097195422978059/1038097349660135474/1060194710485995521),
adding the following instructions to help future contributors avoid
dependency conflicts if they use Conda / Pyenv on their system.

hwchase17 committed:
Harrison/sqlalchemy cache store (#536)

Co-authored-by: Jason Gill <jasongill@gmail.com>

hwchase17 committed:
Harrison/fix and test caching (#538)

hwchase17 committed:
add handling on error (#541)

hwchase17 committed:
Feature: linkcheck-action (#534) (#542)

- Add support for local build and linkchecking of docs
- Add GitHub Action to automatically check links before prior to
publication
- Minor reformat of Contributing readme
- Fix existing broken links

Co-authored-by: Hunter Gerlach <hunter@huntergerlach.com>

Co-authored-by: Hunter Gerlach <HunterGerlach@users.noreply.github.com>
Co-authored-by: Hunter Gerlach <hunter@huntergerlach.com>

HunterGerlach committed:
Add linkcheck badge to signify when/if links are failing (#546)

Detect whether or not most recent GitHub Action running linkcheck was
successful.

hwchase17 committed:
changes to qa chain (#543)

hwchase17 committed:
bump version to 0.0.57 (#548)

hwchase17 committed:
add return_direct flag to tool (#537)

adds a return_direct flag to tools, which just returns the tool output
as the final output

hwchase17 committed:
Harrison/tools priority (#554)

Co-authored-by: Yong723 <50616781+Yongtae723@users.noreply.github.com>

creatorrr committed:
Add finish reason to Generation for usage downstream (#526)

Add `finish_reason` to `Generation` as well as extend
`BaseOpenAI._generate` to include it in the output. This can be useful
for usage in downstream tasks when we need to filter for only
generations that finished because of `"stop"` for example. Maybe we
should add this to `LLMChain` as well?

For more details, see
https://beta.openai.com/docs/guides/completion/best-practices

Signed-off-by: Diwank Singh Tomer <diwank.singh@gmail.com>

hwchase17 committed:
fix map reduce chain (#550)

hwchase17 committed:
RFC: conversational agent (#464)

Co-authored-by: Bruno Bornsztein <bruno.bornsztein@gmail.com>

hwchase17 committed:
fix caching (#555)

BenderV committed:
add openai logit bias (#553)

Add
[`logit_bias`](https://beta.openai.com/docs/api-reference/completions/create#completions/create-logit_bias)
params to OpenAI

See [here](https://beta.openai.com/tokenizer) for the tokenizer.

NB: I see that others (like Cohere) have the same parameter, but since I
don't have an access to it, I don't want to make a mistake.

---

Just to make sure the default "{}" works for openai:
```
from langchain.llms import OpenAI

OPENAI_API_KEY="XXX"

llm = OpenAI(openai_api_key=OPENAI_API_KEY)
llm.generate('Write "test":')

llm = OpenAI(openai_api_key=OPENAI_API_KEY, logit_bias={'9288': -100, '1332': -100, '14402': -100, '6208': -100})
llm.generate('Write "test":')
```

hwchase17 committed:
bump to 0058 (#556)

nickscamara committed:
Docs: side menu to match the order (llms) (#557)

Small quick fix:

Suggest making the order of the menu the same as it is written on the
page (Getting Started -> Key Concepts). Before the menu order was not
the same as it was on the page. Not sure if this is the only place the
menu is affected.

Mismatch is found here:
https://langchain.readthedocs.io/en/latest/modules/llms.html

hwchase17 committed:
map rerank chain (#516)

add a chain that applies a prompt to all inputs and then returns not
only an answer but scores it

add examples for question answering and question answering with sources

hwchase17 committed:
smart text splitter (#530)

smart text splitter that iteratively tries different separators until it
works!

hwchase17 committed:
Harrison/fix splitting (#563)

fix issue where text splitting could possibly create empty docs

hwchase17 committed:
improve documentation on how to pass in custom prompts (#561)

hwchase17 committed:
bump version (#565)

Yongtae723 committed:
fix typo (#570)

I found a typo, which might be important for a conversational Agent.

if My PR is wrong, I am so sorry

Yongtae723 committed:
modify docstring (#569)

Sorry for the detail. this is a correction to the docstring.

hwchase17 committed:
add example selector docs (#564)

hwchase17 committed:
Harrison/deployments (#572)

hwchase17 committed:
gallery updates (#573)

hwchase17 committed:
more gallery (#577)

hwchase17 committed:
Harrison/wolfram alpha (#579)

Co-authored-by: Nicolas <nicolascamara29@gmail.com>

hwchase17 committed:
Harrison/namespace pinecone (#581)

Co-authored-by: mmorzywolek <89693033+mmorzywolek@users.noreply.github.com>

hwchase17 committed:
add similarity score method to faiss (#574)

adds `similarity_search_with_score` to faiss wrapper

hwchase17 committed:
Harrison/combine memories (#582)

Signed-off-by: Diwank Singh Tomer <diwank.singh@gmail.com>
Co-authored-by: Diwank Singh Tomer <diwank.singh@gmail.com>

hwchase17 committed:
Harrison/load from hub (#580)

hwchase17 committed:
bump to version 60 (#583)

hwchase17 committed:
pinecone changes (#590)

Co-authored-by: Smit Shah <who828@gmail.com>
Co-authored-by: iocuydi <46613640+iocuydi@users.noreply.github.com>

hwchase17 committed:
wolfram alpha improvements (#591)

Co-authored-by: Nicolas <nicolascamara29@gmail.com>

hwchase17 committed:
Return source documents option in VectorDBQA (#585) (#592)

Co-authored-by: lesscomfortable <pancho_ingham@hotmail.com>

Co-authored-by: Francisco Ingham <fpingham@gmail.com>
Co-authored-by: lesscomfortable <pancho_ingham@hotmail.com>

hwchase17 committed:
Harrison/update links1 (#594)

update links to be relative

Co-authored-by: Marc Green <marcgreen@users.noreply.github.com>

hwchase17 committed:
add documentation on how to load different chain types (#595)

hwchase17 committed:
bump version to 61 (#596)

eltociear committed:
Fix typo in vector_db_qa.ipynb (#597)

paramter -> parameter

rukmal committed:
Fix minor error in LLM documentation (#602)

samching committed:
Add subsection for colab notebooks (#599)

Motivation is that these don't get lost in the Twitterverse!

babbldev committed:
Added filter argument to pinecone queries, fixes #600 (#601)

Added filter argument to similarity_search() and
similarity_search_with_score()

Co-authored-by: Sam Cartford (MBP) <cartford@hey.com>

mmz-001 committed:
docs: fix typo (#604)

Who828 committed:
[Minor Fix] Fix spacy TextSplitter init (#606)

hwchase17 committed:
add custom prompt for LLMMathChain and SQLDatabase chain (#605)

hwchase17 committed:
bump version to 0.0.62 (#607)

nickscamara committed:
docs: fix small typo (#611)

hwchase17 committed:
expose more serpapi parameters (#609)

fpingham committed:
Custom prompt option for llm_bash and api chains (#612)

Co-authored-by: lesscomfortable <pancho_ingham@hotmail.com>

nickscamara committed:
feat: add custom prompt for QAEvalChain chain (#610)

I originally had only modified the `from_llm` to include the prompt but
I realized that if the prompt keys used on the custom prompt didn't
match the default prompt, it wouldn't work because of how `apply` works.

So I made some changes to the evaluate method to check if the prompt is
the default and if not, it will check if the input keys are the same as
the prompt key and update the inputs appropriately.

Let me know if there is a better way to do this.

Also added the custom prompt to the QA eval notebook.

hwchase17 committed:
improve css (#615)

hwchase17 committed:
bump version to 0063 (#616)

hwchase17 committed:
simplify parsing of the final answer (#621)

hwchase17 committed:
Harrison/wiki update (#622)

Co-authored-by: Rubens Mau <rubensmau@gmail.com>

hwchase17 committed:
more complex sql chain (#619)

add a more complex sql chain that first subsets the necessary tables

hwchase17 committed:
Harrison/new api chain (#623)

Co-authored-by: Francisco Ingham <fpingham@gmail.com>
Co-authored-by: lesscomfortable <pancho_ingham@hotmail.com>

shoelsch committed:
chore: add release workflow (#360)

Adds release workflow that (1) creates a GitHub release and (2)
publishes built artifacts to PyPI

**Release Workflow**
1. Checkout `master` locally and cut a new branch
1. Run `poetry version <rule>` to version bump (e.g., `poetry version
patch`)
1. Commit changes and push to remote branch
1. Ensure all quality check workflows pass
1. Explicitly tag PR with `release` label
1. Merge to mainline

At this point, a release workflow should be triggered because:
* The PR is closed, targeting `master`, and merged
* `pyproject.toml` has been detected as modified
* The PR had a `release` label

The workflow will then proceed to build the artifacts, create a GitHub
release with release notes and uploaded artifacts, and publish to PyPI.

Example Workflow run:
https://github.com/shoelsch/langchain/actions/runs/3711037455/jobs/6291076898
Example Releases: https://github.com/shoelsch/langchain/releases

--

Note, this workflow is looking for the `PYPI_API_TOKEN` secret, so that
will need to be uploaded to the repository secrets. I tested uploading
as far as hitting a permissions issue due to project ownership in Test
PyPI.

hwchase17 committed:
version bump to 0.0.64 (#624)

fpingham committed:
Exclude reference to 'example' in api prompt (#629)

Co-authored-by: lesscomfortable <pancho_ingham@hotmail.com>

francis committed:
fix typo (#636)

there is a small typo in one of the docs.

mmz-001 committed:
Fix CohereError: embed is not an available endpoint on this model (#637)

Running the Cohere embeddings example from the docs:

```python
from langchain.embeddings import CohereEmbeddings
embeddings = CohereEmbeddings(cohere_api_key= cohere_api_key)

text = "This is a test document."
query_result = embeddings.embed_query(text)
doc_result = embeddings.embed_documents([text])
```

I get the error:

```bash
CohereError(message=res['message'], http_status=response.status_code, headers=response.headers)      
cohere.error.CohereError: embed is not an available endpoint on this model
```

This is because the `model` string is set to `medium` which is not
currently available.

From the Cohere docs:

> Currently available models are small and large (default)

hwchase17 committed:
Harrison/cohere experimental (#638)

Co-authored-by: inyourhead <44607279+xettrisomeman@users.noreply.github.com>

rc19 committed:
Fix the env variable for OpenAI Base Url (#639)

For using Azure OpenAI API, we need to set multiple env vars. But as can
be seen in openai package
[here](https://github.com/openai/openai-python/blob/48b69293a3dff649165c1d0e0f9d77d1b52436c5/openai/__init__.py#L35),
the env var for setting base url is named `OPENAI_API_BASE` and not
`OPENAI_API_BASE_URL`. This PR fixes that part in the documentation.

hwchase17 committed:
Harrison/sql db chain (#641)

Co-authored-by: Bruno Bornsztein <bruno.bornsztein@gmail.com>

hwchase17 committed:
Add documentation for custom prompts for Agents (#631) (#640)

- Added a comment interpreting regex for `ZeroShotAgent`
- Added a note to the `Custom Agent` notebook

Co-authored-by: Sam Ching <samuel@duolingo.com>

hwchase17 committed:
add instructions for saving loading (#642)

hwchase17 committed:
bump version to 0065 (#646)

hwchase17 committed:
pinecone docstring (#654)

jmgb27 committed:
typo (#650)

charlesfrye committed:
typo bugfixes in getting started with prompts (#651)

tl;dr: input -> word, output -> antonym, rename to dynamic_prompt
consistently

The provided code in this example doesn't run, because the keys are
`word` and `antonym`, rather than `input` and `output`.

Also, the `ExampleSelector`-based prompt is named `few_shot_prompt` when
defined and `dynamic_prompt` in the follow-up example. The former name
is less descriptive and collides with an earlier example, so I opted for
the latter.

Thanks for making a really cool library!

iocuydi committed:
Add search_kwargs option for VectorDBQAWithSourcesChain (#657)

Allows for passing additional vectorstore params like namespace, etc. to
VectorDBQAWithSourcesChain

Example:
`chain = VectorDBQAWithSourcesChain.from_llm(OpenAI(temperature=0),
vectorstore=store, search_kwargs={"namespace": namespace})`

hwchase17 committed:
Harrison/bing wrapper (#656)

Co-authored-by: Enrico Shippole <henryshippole@gmail.com>

hwchase17 committed:
fix caching (#658)

iocuydi committed:
Add ids parameter for pinecone from_texts / add_texts (#659)

Allow optionally specifying a list of ids for pinecone rather than
having them randomly generated.
This also permits editing the embedding/metadata of existing pinecone
entries, by id.

vertinski committed:
Prompt fix for empty intermediate steps in summarization (#660)

Adding quotation marks around {text} avoids generating empty or
completely random responses from OpenAI davinci-003. Empty or completely
unrelated intermediate responses in summarization messes up the final
result or makes it very inaccurate.
The error from OpenAI would be: "The model predicted a completion that
begins with a stop sequence, resulting in no output. Consider adjusting
your prompt or stop sequences."
This fix corrects the prompting for summarization chain. This works on
API too, the images are for demonstrative purposes.
This approach can be applied to other similar prompts too. 

Examples:

1) Without quotation marks
![Screenshot from 2023-01-20
07-18-19](https://user-images.githubusercontent.com/22897470/213624365-9dfc18f9-5f3f-45d2-abe1-56de67397e22.png)

2) With quotation marks
![Screenshot from 2023-01-20
07-18-35](https://user-images.githubusercontent.com/22897470/213624478-c958e742-a4a7-46fe-a163-eca6326d9dae.png)

hwchase17 committed:
add search kwargs (#664)

hwchase17 committed:
Harrison/quadrant (#665)

Co-authored-by: Kacper Łukawski <kacperlukawski@users.noreply.github.com>

hwchase17 committed:
bump version to 0066 (#667)

trigaten committed:
Create CITATION.cff (#672)

You may want to add doi/orcid

Followed this:
https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-citation-files

hwchase17 committed:
Harrison/load tools kwargs (#681)

Co-authored-by: Bruno Bornsztein <bruno.bornsztein@gmail.com>

frankolson committed:
Update hyperlink in Custom Prompt Template page (#677)

The current link points to a non-existent page. I've updated the link to
match what is on the "Create a custom example selector" page.

<img width="584" alt="Screen Shot 2023-01-21 at 10 33 05 AM"
src="https://user-images.githubusercontent.com/6773706/213879535-d8f2953d-ac37-448d-9b32-fdeb7b73cc32.png">

hwchase17 committed:
strip whitespace (#680)

hwchase17 committed:
Harrison/few shot yaml (#682)

Co-authored-by: vintro <77507980+vintrocode@users.noreply.github.com>

rvorias committed:
add faiss local saving/loading (#676)

- This uses the faiss built-in `write_index` and `load_index` to save
and load faiss indexes locally
- Also fixes #674
- The save/load functions also use the faiss library, so I refactored
the dependency into a function

eltociear committed:
Fix typo in hyde.ipynb (#688)

therefor -> therefore

sjwhitmore committed:
ConversationEntityMemory: Chain which uses an entity extraction & sum… (#678)

…marization prompt to maintain a key-value store of memory information

cc @devennavani

Co-authored-by: Harrison Chase <hw.chase.17@gmail.com>

hwchase17 committed:
bump version to 0.0.67 (#689)

hwchase17 committed:
verbose flag (#683)

nickscamara committed:
docs: small typo google_search.md (#692)

nickscamara committed:
docs: small typo on serpapi.md (#693)

delip commented on langchain/llms/accelerate.py:
```diff
@@ -0,0 +1,148 @@
+
+        # what else models are available and compatible? TODO
+
+        weights_path = "./accelerated_model_weights/" + model_name
```
Also, why is the weights path root hardcoded to current directory "."?

delip commented on langchain/llms/accelerate.py:
```diff
@@ -0,0 +1,148 @@
+        # what else models are available and compatible? TODO
+
+        weights_path = "./accelerated_model_weights/" + model_name
+        if not os.path.exists(weights_path):
```
Doesn't huggingface automatically do this test when you call `from_pretrained`?

delip commented on langchain/llms/accelerate.py:
```diff
@@ -0,0 +1,148 @@
+            model.save_pretrained(weights_path)
+        config = AutoConfig.from_pretrained(model_name)
+
+        config.use_cache = False
```
why disable HF cache use and create a new cache path manually as you are doing here?

delip commented:
> nice! im not super familiar with accerlated stuff. how generic is the get_accelerated_gpt code?

@hwchase17 Since I ran into similar issue and I am familiar with this, I left a bunch of comments to @thepok in a review. 

@thepok Besides the review comments, can you add an example of how Accelarate is used (preferably not as a notebook for review purposes)?

SauravMaheshkar committed:
chore: move coverage config to pyproject (#694)

This PR aims to move the contents of `.coveragerc` to `pyproject.toml`
to make the overall file structure more minimal.

hwchase17 committed:
Harrison/serpapi extra tools (#691)

Co-authored-by: Bruno Bornsztein <bruno.bornsztein@gmail.com>

amosjyng committed:
Update examples to fix execution problems (#685)

On the [Getting Started
page](https://langchain.readthedocs.io/en/latest/modules/prompts/getting_started.html)
for prompt templates, I believe the very last example

```python
print(dynamic_prompt.format(adjective=long_string))
```

should actually be

```python
print(dynamic_prompt.format(input=long_string))
```

The existing example produces `KeyError: 'input'` as expected

***

On the [Create a custom prompt
template](https://langchain.readthedocs.io/en/latest/modules/prompts/examples/custom_prompt_template.html#id1)
page, I believe the line

```python
Function Name: {kwargs["function_name"]}
```

should actually be

```python
Function Name: {kwargs["function_name"].__name__}
```

The existing example produces the prompt:

```
        Given the function name and source code, generate an English language explanation of the function.
        Function Name: <function get_source_code at 0x7f907bc0e0e0>
        Source Code:
        def get_source_code(function_name):
    # Get the source code of the function
    return inspect.getsource(function_name)

        Explanation:
```

***

On the [Example
Selectors](https://langchain.readthedocs.io/en/latest/modules/prompts/examples/example_selectors.html)
page, the first example does not define `example_prompt`, which is also
subtly different from previous example prompts used. For user
convenience, I suggest including

```python
example_prompt = PromptTemplate(
    input_variables=["input", "output"],
    template="Input: {input}\nOutput: {output}",
)
```

in the code to be copy-pasted

hwchase17 committed:
fix whitespace for conversational agent (#690)

hwchase17 committed:
Harrison/openai callback (#684)

xloem committed:
HuggingFacePipeline: Forward model_kwargs. (#696)

Since the tokenizer and model are constructed manually, model_kwargs
needs to
be passed to their constructors. Additionally, the pipeline has a
specific
named parameter to pass these with, which can provide forward
compatibility if
they are used for something other than tokenizer or model construction.

scadEfUr committed:
added common prompt load method (#699)

Co-authored-by: scadEfUr

thepok commented:
@delip i have no interest at the moment to implement this further. Feel free to implement your suggestions or it could be possible to merge it in the huggingface llm loader

hwchase17 committed:
version 0068 (#701)

scottleibrand committed:
remove legacy embedding model name (#703)

Now that OpenAI has deprecated all embeddings models except
text-embedding-ada-002, we should stop specifying a legacy embedding
model in the example. This will also avoid confusion from people (like
me) trying to specify model="text-embedding-ada-002" and having that
erroneously expanded to text-search-text-embedding-ada-002-query-001

hwchase17 committed:
langchain hub docs (#704)

Co-authored-by: scadEfUr <123224380+scadEfUr@users.noreply.github.com>

nickscamara committed:
docs: Update vector_db_qa_with_sources.ipynb (#706)

feynmanliang committed:
Fix typo (#705)

hwchase17 committed:
bump version to 0069 (#710)

hwchase17 committed:
clarify use cases (#711)

kacperlukawski committed:
Hotfix: Qdrant.from_text embeddings (#713)

I'm providing a hotfix for Qdrant integration. Calculating a single
embedding to obtain the vector size was great idea. However, that change
introduced a bug trying to put only that single embedding into the
database. It's fixed. Right now all the embeddings will be pushed to
Qdrant.

kacperlukawski committed:
Convert numpy arrays to lists in HuggingFaceEmbeddings (#714)

`SentenceTransformer` returns a NumPy array, not a `List[List[float]]`
or `List[float]` as specified in the interface of `Embeddings`. That PR
makes it consistent with the interface.

feynmanliang committed:
Add namespace to Pinecone.from_index (#716)

Resolves https://github.com/hwchase17/langchain/issues/718

shanginn committed:
Fix not imported 'validator' (#715)

otherwise `@validator("input_variables")` do not work

I-E-E-E committed:
fix a url (#719)

samheutmaker committed:
fix typos in readme and text splitter docs (#720)

Fix typos in readme and TextSplitter documentation.

hwchase17 committed:
Harrison/serialize llm chain (#671)

scadEfUr committed:
move hyde into chains (#728)

Co-authored-by: scadEfUr <>

hwchase17 committed:
add stop to stream (#729)

hwchase17 committed:
Harrison/generate blog post (#732)

Co-authored-by: Ren <yirenlu92@users.noreply.github.com>

amosjyng committed:
Fix sqlalchemy warnings when running tests (#733)

This has been bugging me when running my own tests that call langchain
methods :P

fpingham committed:
Hotfix: distance_func and collection_name must not be in kwargs (#735)

If `distance_func` and `collection_name` are in `kwargs` they are sent
to the `QdrantClient` which results in an error being raised.

Co-authored-by: Francisco Ingham <>

thepok committed:
Merge branch 'accelerate' of https://github.com/thepok/langchain into accelerate

dev2049 commented:
stale

Bec-k commented:
ugh, so this have not been integrated?

thepok commented:
nope, but i think huggingfaces transformer lib can it now by itself, and hugginfaces transformer is already in langchain :)
